### PR TITLE
feat: add new skaffold verify command and verify phase

### PR DESCRIPTION
**Pull request by aaron-prindle**

### CODE REVIEW

**Comment by aaron-prindle** [comment]
Failing kokoro test is a known flake, tests are passing

File: pkg/skaffold/runner/v2/verify.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		return err
+	}
+
+	if !r.runCtx.Opts.IterativeStatusCheck {
+		// run final aggregated status check only if iterative status check is turned off.
+		if err = r.verifier.GetStatusMonitor().Check(ctx, statusCheckOut); err != nil {
+			eventV2.TaskFailed(constants.Verify, err)
+			return err
+		}
+	}
```
Comment (tejal29) [inline pkg/skaffold/runner/v2/verify.go:127]: Not sure I understand this.  Why does the verifier depend on the iterative statusCheck?

Would the verifier not run if iterative status check is on

File: pkg/skaffold/runner/verifier.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	for _, p := range runCtx.GetPipelines() {
+		d, err := docker.NewVerifier(ctx, runCtx, labeller, p.Verify, runCtx.PortForwardResources(), runCtx.VerifyDockerNetwork())
+		if err != nil {
+			return nil, err
+		}
+		deployers = append(deployers, d)
+	}
+
+	return verify.NewDeployerMux(deployers, runCtx.IterativeStatusCheck()), nil
```
Comment (tejal29) [inline pkg/skaffold/runner/verifier.go:49]: should this be NewVerifierMux?
```suggestion
	return verify.NewVerifierMux(deployers, runCtx.IterativeStatusCheck()), nil
```

File: pkg/skaffold/verify/docker/deploy.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/log"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/output"
+	olog "github.com/GoogleContainerTools/skaffold/pkg/skaffold/output/log"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/status"
+	pkgsync "github.com/GoogleContainerTools/skaffold/pkg/skaffold/sync"
+)
+
+type Deployer struct {
```
Comment (tejal29) [inline pkg/skaffold/verify/docker/deploy.go:55]: Is Verifier a Deployer?
Or shd we rename this to `Verifyier`?

File: pkg/skaffold/verify/docker/deploy.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+// TrackContainerFromBuild adds an artifact and its newly-associated container
+// to the container tracker.
+func (d *Deployer) TrackContainerFromBuild(artifact graph.Artifact, container tracker.Container) {
+	d.tracker.Add(artifact, container)
+}
+
+// Deploy deploys built artifacts by creating containers in the local docker daemon
+// from each artifact's image.
+func (d *Deployer) Deploy(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) error {
+	var err error
```
Comment (tejal29) [inline pkg/skaffold/verify/docker/deploy.go:126]:
```suggestion
func (d *Deployer) Run(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) error {
	var err error
```

File: pkg/skaffold/verify/docker/deploy.go [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+			return d.deploy(ctx, out, nb, *testCase)
+		})
+	}
+	d.TrackBuildArtifacts(builds)
+	return s.Wait()
+}
+
+// deploy creates a container in the local docker daemon from a build artifact's image.
+func (d *Deployer) deploy(ctx context.Context, out io.Writer, artifact graph.Artifact, tc latest.VerifyTestCase) error {
+	out, ctx = output.WithEventContext(ctx, out, constants.Verify, tc.Name)
```
Comment (tejal29) [inline pkg/skaffold/verify/docker/deploy.go:170]:
```suggestion
func (d *Deployer) createContainer(ctx context.Context, out io.Writer, artifact graph.Artifact, tc latest.VerifyTestCase) error {
	out, ctx = output.WithEventContext(ctx, out, constants.Verify, tc.Name)
```

File: pkg/skaffold/verify/docker/deploy.go [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+// setupDebugging configures the provided artifact's image for debugging (if applicable).
+// The provided container configuration receives any relevant modifications (e.g. ENTRYPOINT, CMD),
+// and any init containers for populating the shared debug volume will be created.
+// A list of port bindings for the exposed debuggers is returned to be processed alongside other port
+// forwarding resources.
+func (d *Deployer) setupDebugging(ctx context.Context, out io.Writer, artifact graph.Artifact, containerCfg *container.Config) (nat.PortMap, error) {
+	initContainers, err := d.debugger.TransformImage(ctx, artifact, containerCfg)
+	if err != nil {
```
Comment (tejal29) [inline pkg/skaffold/verify/docker/deploy.go:242]: Does this allow users to debug tests?
I am not sure if this was listed in the design document. How would user's be able to use this in CLI or IDE?

File: pkg/skaffold/verify/docker/deploy.go [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+	return d.syncer
+}
+
+func (d *Deployer) GetStatusMonitor() status.Monitor {
+	return d.monitor
+}
+
+func (d *Deployer) RegisterLocalImages([]graph.Artifact) {
+	// all images are local, so this is a noop
+}
```
Comment (tejal29) [inline pkg/skaffold/verify/docker/deploy.go:395]: Do we need this for verifier?

File: pkg/skaffold/verify/verify.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/log"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/status"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/sync"
+)
+
+// Deployer is the Deploy API of skaffold and responsible for deploying
+// the build results to a Kubernetes cluster
+type Deployer interface {
+	// Deploy should ensure that the build results are deployed to the Kubernetes
+	// cluster.
```
Comment (tejal29) [inline pkg/skaffold/verify/verify.go:35]: Should this be verifier?

**Review by tejal29 — COMMENTED** [review]
Few naming conventions commits.

Do we need the verifier interface to be same as `Deployer` interface?

Happy to chat in person

File: pkg/skaffold/runner/v2/verify.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+	// Logs should be retrieved up to just before the deploys
+	r.verifier.GetLogger().SetSince(time.Now())
+
+	// Start printing the logs after deploy is finished
+	if err := r.verifier.GetLogger().Start(ctx, out); err != nil {
+		return fmt.Errorf("starting logger: %w", err)
+	}
+
+	// First deploy
```
[the remark this answers is not in the document]
Comment (aaron-prindle) [reply pkg/skaffold/runner/v2/verify.go]: I've updated the comments here to reflect this is verify

**Review by tejal29 — APPROVED** [review]

**Pull request merged**